tuned/plugins/plugin_scheduler.py

```diff
@@ -350,8 +410,103 @@ def _convert_sched_cfg(self, vals):
 		affinity = self._convert_affinity(affinity)
 		return (rule_prio, scheduler, priority, affinity, regex)
 
+	def _cgroup_create_group(self, cgroup):
+		path = "%s/%s" % (self._cgroup_mount_point, cgroup)
+		try:
+			os.mkdir(path, consts.DEF_CGROUP_MODE)
+		except OSError as e:
+			log.error("Unable to create cgroup '%s': %s" % (path, e))
+		if (not self._cmd.write_to_file("%s/%s" % (path, "cpuset.mems"),
+				self._cmd.read_file("%s/%s" % (self._cgroup_mount_point, "cpuset.mems"), no_error = True),
+				no_error = True)):
+					log.error("Unable to initialize 'cpuset.mems ' for cgroup '%s'" % path)
+		if (not self._cmd.write_to_file("%s/%s" % (path, "cpuset.cpus"),
+				self._cmd.read_file("%s/%s" % (self._cgroup_mount_point, "cpuset.cpus"), no_error = True),
+				no_error = True)):
+					log.error("Unable to initialize 'cpuset.cpus' for cgroup '%s'" % path)
+
+	def _cgroup_initialize_groups(self):
+		if self._cgroup is not None and not self._cgroup in self._cgroups:
+			self._cgroup_create_group(self._cgroup)
+		for cg in self._cgroups:
+			self._cgroup_create_group(cg)
+
+	def _cgroup_initialize(self):
+		log.debug("Initializing cgroups settings")
+		try:
+			os.makedirs(self._cgroup_mount_point, consts.DEF_CGROUP_MODE)
+		except OSError as e:
+			log.error("Unable to create cgroup mount point: %s" % e)
+		(ret, out) = self._cmd.execute(["mount", "-t", "cgroup", "-o", "cpuset", "cpuset", self._cgroup_mount_point])
+		if ret != 0:
+			log.error("Unable to mount '%s'" % self._cgroup_mount_point)
+
+	def _remove_dir(self, cgroup):
+		try:
+			os.rmdir(cgroup)
+		except OSError as e:
+			log.error("Unable to remove directory '%s': %s" % (cgroup, e))
+
+	def _cgroup_finalize_groups(self):
+		if self._cgroup is not None and not self._cgroup in self._cgroups:
+			self._remove_dir("%s/%s" % (self._cgroup_mount_point, self._cgroup))
+		for cg in self._cgroups:
+			self._remove_dir("%s/%s" % (self._cgroup_mount_point, cg))
+
+	def _cgroup_finalize(self):
+		log.debug("Removing cgroups settings")
+		(ret, out) = self._cmd.execute(["umount", self._cgroup_mount_point])
+		if ret != 0:
+			log.error("Unable to umount '%s'" % self._cgroup_mount_point)
+			return False
+		self._remove_dir(self._cgroup_mount_point)
+		d = os.path.dirname(self._cgroup_mount_point)
+		if (d != "/"):
+			self._remove_dir(d)
+
+	def _cgroup_set_affinity_one(self, cgroup, affinity, backup = False):
+		log.debug("Setting cgroup '%s' affinity to '%s'" % (cgroup, affinity))
+		path = "%s/%s/%s" % (self._cgroup_mount_point, cgroup, "cpuset.cpus")
+		if backup:
+			orig_affinity = self._cmd.read_file(path, err_ret = "ERR", no_error = True)
+			if orig_affinity != "ERR":
+				self._cgroups_original_affinity[cgroup] = orig_affinity
+			else:
+				log.warn("Unable to save affinity of cgroup '%s'" % cgroup)
+		if not self._cmd.write_to_file(path, affinity, no_error = True):
```

Comment:
Should we change the affinity if we failed to read the original affinity? In general I think we shouldn't change anything that we can't change back. There's an example of this in `_tune_process_affinity` (the "Refusing to set..." error).

Reply:
I am not sure, from the performance point of view I would prefer to set it (or attempt to set it), because I would care about the performance, not about the possibility to roll-back. But I agree we should be consistent in the behaviour. Also this is a bit different than just the plain affinity, because in case of removal of the cgroup we move the task to the root cgroup. But I dont' have problem to change it.

Reply:
And if the failure is not really expected, I think `log.error` is more appropriate than `log.warn`.

Reply:
> Well, it's not really _expected_ for the read operation to fail, is it? If it does fail, I think it either means that there's a bug in Tuned (in which case it's safer to do nothing), or there's some deeper problem in the system. I think a failure of the read operation is mostly hypothetical, so we should prefer consistency over hypothetical performance gains.
>
It's the read of kernel special file, not regular file, which is handled by the cgroup kernel driver. I am not familiar with it's implementation, so the failure can mean nearly anything.

> > Also this is a bit different than just the plain affinity, because in case of removal of the cgroup we move the task to the root cgroup. But I dont' have problem to change it.
> 
> Well, in the case where the cgroups are removed, there's not really a point in doing the backup in the first place. (Or at least that's my current understanding.) So in that case, it would be better not to do the backup at all.
> 
The cgroup remove operation could fail and the cgroup could also exists before Tuned attempted to create it, in such case it's safer to backup.

> What we should be talking about here is the case where the cgroups are _not_ removed, i.e. they are user provided. In that case, we're not really talking about restoring affinity of the _tasks_, but about restoring affinity of the _cgroups_. And we should treat the cgroups as any other thing that deserves rollback.

As I wrote earlier I am for consistency in behaviour, so I am not against your proposal.

Reply:
> > Well, it's not really _expected_ for the read operation to fail, is it? If it does fail, I think it either means that there's a bug in Tuned (in which case it's safer to do nothing), or there's some deeper problem in the system. I think a failure of the read operation is mostly hypothetical, so we should prefer consistency over hypothetical performance gains.
> 
> It's the read of kernel special file, not regular file, which is handled by the cgroup kernel driver. I am not familiar with it's implementation, so the failure can mean nearly anything.

Fair enough. However it seems like a very basic operation that I really would not expect to fail.

> > > Also this is a bit different than just the plain affinity, because in case of removal of the cgroup we move the task to the root cgroup. But I dont' have problem to change it.
> > 
> > 
> > Well, in the case where the cgroups are removed, there's not really a point in doing the backup in the first place. (Or at least that's my current understanding.) So in that case, it would be better not to do the backup at all.
> 
> The cgroup remove operation could fail and the cgroup could also exists before Tuned attempted to create it, in such case it's safer to backup.

I'd suggest that if creating the cgroup fails because the cgroup is already there, then we should abort everything and not touch those cgroups at all :). Because that's clearly either a bug or misconfiguration, so it's safer not to do anything.

So no cgroup touching => no backup needed :).

Reply:
This wouldn't be good for recovery, e.g. recovery from deamon crash etc. In such case just the restart (or double restart) fixes the problem.

Reply:
For those cases, there probably should be a `force` option to make Tuned ignore the directory creation error. Or perhaps a way to tell Tuned to cleanup the leftover directories.

Reply:
This has to be engine / tuned-adm wide.

Reply:
To make it easy to use (i.e. without modifying the profile), yes, I think it would have to be a global option. But I think it's the right solution.

Reply:
Note that I'm not saying that you should now go off and implement the force/cleanup option. We can do that later :).

I guess the only thing I'd like to do right now is to decide the general direction where the code is going to go.